File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Map<Integer, Integer> lastIndex = new HashMap<>();

        for (int r = 0; r < nums.length; r++) {
            int val = nums[r];

            if (lastIndex.containsKey(val)) {
                int l = lastIndex.get(val);
                if (r - l <= k) {
                    return true;
                }
            }
            lastIndex.put(val, r);
        }

        return false;
    }
}

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        
        l = 0
        n = len(nums)

        seen = set()

        for r in range(n):
            x = nums[r]
            if r - l > k:
                seen.remove(nums[l])
                l += 1
            
            if x in seen:
                return True
            seen.add(x)
        
        return False
